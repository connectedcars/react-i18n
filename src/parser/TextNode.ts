import { SyntaxKind, DocNode } from './types'

export class TextNode {
  public kind = SyntaxKind.text
  public text: string
  public parent: DocNode | null = null

  constructor(text: string, parent: DocNode | null) {
    this.text = text
    this.parent = parent
  }
}
