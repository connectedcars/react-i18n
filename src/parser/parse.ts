import { ElementNode } from './ElementNode'
import { TextNode } from './TextNode'
import { DocNode, SyntaxKind, Token } from './types'

const addToken = (
  stack: Token[],
  line: string,
  isTextNode: boolean,
  patternString: string
): void => {
  if (!line) {
    return
  }

  if (isTextNode) {
    stack.push({
      kind: SyntaxKind.text,
      text: line,
    })
    return
  }

  let match: RegExpMatchArray | null

  // <tag attrs />
  if ((match = line.match(/<([^\s<>]+) ?([^<>]*)\/>/))) {
    stack.push({
      kind: SyntaxKind.self,
      tagName: match[1],
      attributes: match[2],
      text: line,
    })
    return
  }

  // </tag>
  if ((match = line.match(/<\/([^\s<>]+)>/))) {
    stack.push({
      kind: SyntaxKind.close,
      tagName: match[1],
      text: line,
    })
    return
  }

  // <tag attrs>
  if ((match = line.match(/<([^\s<>]+) ?([^<>]*)>/))) {
    stack.push({
      kind: SyntaxKind.open,
      tagName: match[1],
      attributes: match[2],
      text: line,
    })
    return
  }

  // {text}
  if ((match = line.match(new RegExp(patternString)))) {
    stack.push({
      kind: SyntaxKind.self,
      tagName: match[1],
      attributes: match[1],
      text: line,
    })
    return
  }

  return
}

const tokenize = (str: string, patternString: string): Token[] => {
  const pattern = new RegExp(
    `(<[^<>]+>${patternString ? `|${patternString}` : ''})`,
    'g'
  )
  const stack: Token[] = []

  let match: RegExpExecArray | null = null
  let pointer: number = 0
  while ((match = pattern.exec(str))) {
    addToken(stack, str.slice(pointer, match.index), true, patternString)
    addToken(stack, match[1], false, patternString)
    pointer = match.index + match[0].length
  }

  const hasMoreText = pointer < str.length
  if (hasMoreText) {
    addToken(stack, str.slice(pointer), true, patternString)
  }

  return stack
}

const lex = (tokens: Token[]): DocNode[] => {
  const children: DocNode[] = []

  if (tokens.length === 0) {
    return children
  }

  let currentNode: DocNode | null = null
  for (const node of tokens) {
    if (node.kind === SyntaxKind.text) {
      if (currentNode instanceof ElementNode) {
        currentNode.children.push(new TextNode(node.text, currentNode))
      } else {
        children.push(new TextNode(node.text, null))
      }

      continue
    }

    if (node.kind === SyntaxKind.close) {
      if (!(currentNode instanceof ElementNode)) {
        throw new Error(`missing open node: '${node.tagName}'`)
      }
      if (node.tagName !== currentNode.tagName) {
        throw new Error(`missing close node: '${currentNode.tagName}'`)
      }

      currentNode = currentNode.parent
      continue
    }

    if (node.kind === SyntaxKind.open) {
      const newNode = new ElementNode(node, currentNode)
      if (currentNode instanceof ElementNode) {
        currentNode.children.push(newNode)
      } else {
        children.push(newNode)
      }

      currentNode = newNode
      continue
    }

    if (node.kind === SyntaxKind.self) {
      if (currentNode instanceof ElementNode) {
        currentNode.children.push(new ElementNode(node, currentNode))
      } else {
        children.push(new ElementNode(node, null))
      }
      continue
    }
  }

  if (
    currentNode instanceof ElementNode &&
    currentNode.kind === SyntaxKind.open
  ) {
    throw new Error(`missing close node: '${currentNode.tagName}'`)
  }

  return children
}

const parse = (
  str: string,
  patternString: string = '{([\\w-]+)}'
): DocNode[] => {
  // Break our string into tokens
  const lines = tokenize(str, patternString)
  // Add extra context to our tokens and validate the syntax
  const ast = lex(lines)
  // Return our syntax tree
  return ast
}

export { parse }
