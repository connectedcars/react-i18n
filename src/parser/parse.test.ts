import parse from "./parse";
import TextNode from "./TextNode";

describe('parse', () => {
  it('can parse strings', () => {
    expect(parse('foo <foo>bar</foo> {bar}')).toMatchSnapshot()
    expect(parse('foobar')).toMatchSnapshot()
    expect(parse('foobar <bar /><bar/><bar attributes />')).toMatchSnapshot()
    expect(parse('foobar <bar attributes />')).toMatchSnapshot()
    expect(parse('foobar {bar} <bar />')).toMatchSnapshot()
  })

  it('throws an error when missing open nodes', () => {
    expect(() => parse('bar</a>')).toThrowError('missing open node: a')
    expect(() => parse('<a>bar</a></b>')).toThrowError('missing open node: b')
    expect(() => parse('<a>bar</a></a>')).toThrowError('missing open node: a')
  })

  it('throws an error when missing close nodes', () => {
    expect(() => parse('<a>bar')).toThrowError('missing close node: a')
    expect(() => parse('<b><a>bar')).toThrowError('missing close node: a')
    expect(() => parse('<b><a>bar</a>')).toThrowError('missing close node: b')
  })
})
