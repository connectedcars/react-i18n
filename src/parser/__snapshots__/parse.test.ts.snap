// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse can parse strings 1`] = `
[
  TextNode {
    "kind": "text",
    "parent": null,
    "text": "foo ",
  },
  ElementNode {
    "attributes": "",
    "children": [
      TextNode {
        "kind": "text",
        "parent": [Circular],
        "text": "bar",
      },
    ],
    "kind": "open",
    "parent": null,
    "tagName": "foo",
    "text": "<foo>",
  },
  TextNode {
    "kind": "text",
    "parent": null,
    "text": " ",
  },
  ElementNode {
    "attributes": "bar",
    "children": [],
    "kind": "self",
    "parent": null,
    "tagName": "bar",
    "text": "{bar}",
  },
]
`;

exports[`parse can parse strings 2`] = `
[
  TextNode {
    "kind": "text",
    "parent": null,
    "text": "foobar",
  },
]
`;

exports[`parse can parse strings 3`] = `
[
  TextNode {
    "kind": "text",
    "parent": null,
    "text": "foobar ",
  },
  ElementNode {
    "attributes": "",
    "children": [],
    "kind": "self",
    "parent": null,
    "tagName": "bar",
    "text": "<bar />",
  },
  ElementNode {
    "attributes": "",
    "children": [],
    "kind": "self",
    "parent": null,
    "tagName": "bar",
    "text": "<bar/>",
  },
  ElementNode {
    "attributes": "attributes ",
    "children": [],
    "kind": "self",
    "parent": null,
    "tagName": "bar",
    "text": "<bar attributes />",
  },
]
`;

exports[`parse can parse strings 4`] = `
[
  TextNode {
    "kind": "text",
    "parent": null,
    "text": "foobar ",
  },
  ElementNode {
    "attributes": "attributes ",
    "children": [],
    "kind": "self",
    "parent": null,
    "tagName": "bar",
    "text": "<bar attributes />",
  },
]
`;

exports[`parse can parse strings 5`] = `
[
  TextNode {
    "kind": "text",
    "parent": null,
    "text": "foobar ",
  },
  ElementNode {
    "attributes": "bar",
    "children": [],
    "kind": "self",
    "parent": null,
    "tagName": "bar",
    "text": "{bar}",
  },
  TextNode {
    "kind": "text",
    "parent": null,
    "text": " ",
  },
  ElementNode {
    "attributes": "",
    "children": [],
    "kind": "self",
    "parent": null,
    "tagName": "bar",
    "text": "<bar />",
  },
]
`;

exports[`parse throws an error when missing close nodes 1`] = `"missing close node: 'a'"`;

exports[`parse throws an error when missing close nodes 2`] = `"missing close node: 'a'"`;

exports[`parse throws an error when missing close nodes 3`] = `"missing close node: 'b'"`;

exports[`parse throws an error when missing open nodes 1`] = `"missing open node: 'a'"`;

exports[`parse throws an error when missing open nodes 2`] = `"missing open node: 'b'"`;

exports[`parse throws an error when missing open nodes 3`] = `"missing open node: 'a'"`;
